records/chains.go

```diff
@@ -0,0 +1,80 @@
+package records
+
+import (
+	"github.com/mesosphere/mesos-dns/records/labels"
+)
+
+// chain is a generation func that consumes record-like strings and does
+// something with them
+type chain func(...string)
+
+const (
+	protocolNone = "" // for readability
+	domainNone   = "" // for readability
+)
+
+// withProtocol appends `._{protocol}.{framework}` to records. if protocol is "" then
+// the protocols "tcp" and "udp" are assumed.
+func withProtocol(protocol, framework string, spec labels.Func, gen chain) chain {
+	return func(records ...string) {
+		protocol = spec(protocol)
+		if protocol != protocolNone {
+			for i := range records {
+				records[i] += "._" + protocol + "." + framework
+			}
+		} else {
+			records = append(records, records...)
```

Comment:
Is there any reason to reuse records, and do this in place? Does Golang actually care that much or optimize that much do this in place? Can we just create another slice, and populate that? I think that would make this code much easier to read.